Coalesce daskhub into basehub

yuvipanda opened the pull request:
I have timeboxed myself to 30min, and I have verified that this gives us a fairly seamless migration pathway for existing installations of daskhub (no re-install needed), and allows us to mix multiple dependencies (primarily binderhub-service and dask-gateway) more easily. It does this by taking the following two actions:

- Make dask-gateway a conditional dependency of basehub, rather than daskhub
- Move all remaining daskhub/values.yaml config into basehub, either under dask-gateway, or under an extraConfig that's enabled conditionally (via `custom.daskGateway.enabled`)

The end result here is that `daskhub` is now a stub that is *completely empty*, and simply passes through to basehub! So basehub is where *all* the action is, and `daskhub` is a simple backwards compatible passthrough.

I've tested this on the openscapes staging hub and verified that everything works as it should! I only had to make the following three config additions:

1. Set `basehub.dask-gateway.enabled` to `true`
2. Set `basehub.jupyterhub.custom.daskGateway.enabled` to `true`.
3. Set `basehub.jupyterhub.singleuser.cloudMetadata.blockWithIptables` to `true`. We may not need this, but I didn't want to investigate that in this spike.

Remaining things to do to complete this migration:

- [x] Remove *all* references to `daskhub` in our deployer code. If something needs to know if dask-gateway is enabled, it should check for `basehub.dask-gateway.enabled`
- [x] Make appropriate config changes to all existing daskhub installations so this is a seamless transition.
- [x] Document the fact that some of our existing hub config will have a `daskhub` moniker as a historical artifact. I think the additional minor cognitive overhead is totally worth the fact that we won't have to re-install any hubs!

Ref https://github.com/2i2c-org/infrastructure/issues/4087

GeorgianaElena reviewed (commented):
This looks good to me @yuvipanda. ~There's one thing that I'm unsure if was intentionally left of or I just don't see it, which is these dask specific env vars set on the user pods:~

~`https://github.com/2i2c-org/infrastructure/blob/204b48ac78cada6ecec74bf3460a15bd99d3a9a0/helm-charts/daskhub/values.yaml#L18-L52`~

UPDATE:

Found them https://github.com/2i2c-org/infrastructure/blob/fc5778975ef48e67dbdb2f083c391415b333f574/helm-charts/basehub/values.yaml#L1592-L1597 🚀

GeorgianaElena commented on config/clusters/openscapes/staging.values.yaml:
```diff
@@ -1,8 +1,13 @@
       eks.amazonaws.com/role-arn: arn:aws:iam::783616723547:role/openscapeshub-staging
   jupyterhub:
+    custom:
+      daskGateway:
+        enabled: true
```
Please feel free to dismiss this comment as I don't feel very strongly about it, but I cannot seem to decide if having these two be named the similar (`dask-gateway` and `daskGateway`) is confusing or it's good because it "ties" them together 😄 

I think personally I'd name `jupyterhub.custom.daskGateway` `jupyterhub.custom.daskhub` `jupyterhub.custom.daskhubType` for an easier transition from what we have. But then again I'm bad with names so please feel free to ignore.

Also, I think it's worth adding them to the schema definition for reference although not mandatory.

yuvipanda replied:
Ideally I'd love for you to pick up this and finish it, and you should name them as you see fit!

yuvipanda replied:
I think it would be useful for it to be consistent with everything else under `jupyterhub.custom`, whatever that is.

GeorgianaElena commented on config/clusters/openscapes/staging.values.yaml:
```diff
@@ -1,8 +1,13 @@
 basehub:
```
I think this can now go and together with it all the hardcoded special-cases we have all over the place 😅 yuhuuu 🚀

GeorgianaElena commented:
@yuvipanda, I believe this is now ready to go. I would really appreciate a quick look over the PR and the docs piece to make sure we are on the same page.

Also, if you have any ideas on how I could check this is working as it should before merging that would be great. So far I did a deploy on the 2i2c dask-staging hub and ran the hub health checks. Is this enough? Thank you.

yuvipanda commented:
Here's the code I'm using to test this:

```python
from dask_gateway import Gateway

gateway = Gateway()

cluster = gateway.new_cluster()
cluster.scale(2)

client = cluster.get_client()

import dask.array as da

a = da.random.normal(size=(1000, 1000), chunks=(500, 500))

a.mean().compute()
```

When run, you need to verify two things:

1. Two `dask-worker` pods came up
2. You get an answer (some number) at the end.

This might take a couple minutes to execute, especially if new nodes need to come up for it.

yuvipanda commented:
I tested by running this with `--debug`. Here's the set of things it changed:

```
history.go:56: [debug] getting history for release staging
upgrade.go:155: [debug] preparing upgrade for staging
upgrade.go:163: [debug] performing update for staging
upgrade.go:356: [debug] creating upgraded release for staging
client.go:486: [debug] Starting delete for "nfs-share-creator" Job
client.go:490: [debug] Ignoring delete failure for "nfs-share-creator" batch/v1, Kind=Job: jobs.batch "nfs-share-creator" not found
wait.go:66: [debug] beginning wait for 1 resources to be deleted with timeout of 5m0s
client.go:142: [debug] creating 1 resource(s)
client.go:712: [debug] Watching for changes to Job nfs-share-creator with timeout of 5m0s
client.go:740: [debug] Add/Modify event for nfs-share-creator: ADDED
client.go:779: [debug] nfs-share-creator: Jobs active: 0, jobs failed: 0, jobs succeeded: 0
client.go:740: [debug] Add/Modify event for nfs-share-creator: MODIFIED
client.go:779: [debug] nfs-share-creator: Jobs active: 1, jobs failed: 0, jobs succeeded: 0
client.go:740: [debug] Add/Modify event for nfs-share-creator: MODIFIED
client.go:779: [debug] nfs-share-creator: Jobs active: 0, jobs failed: 0, jobs succeeded: 0
client.go:740: [debug] Add/Modify event for nfs-share-creator: MODIFIED
client.go:486: [debug] Starting delete for "nfs-share-creator" Job
wait.go:66: [debug] beginning wait for 1 resources to be deleted with timeout of 5m0s
client.go:393: [debug] checking 53 resources for changes
client.go:693: [debug] Patch PriorityClass "staging-default-priority" in namespace 
client.go:693: [debug] Patch PriorityClass "staging-user-placeholder-priority" in namespace 
client.go:693: [debug] Patch NetworkPolicy "hub" in namespace staging
client.go:693: [debug] Patch NetworkPolicy "proxy" in namespace staging
client.go:693: [debug] Patch NetworkPolicy "singleuser" in namespace staging
client.go:693: [debug] Patch PodDisruptionBudget "user-placeholder" in namespace staging
client.go:693: [debug] Patch PodDisruptionBudget "user-scheduler" in namespace staging
client.go:684: [debug] Looks like there are no changes for ServiceAccount "controller-staging-dask-gateway"
client.go:684: [debug] Looks like there are no changes for ServiceAccount "api-staging-dask-gateway"
client.go:684: [debug] Looks like there are no changes for ServiceAccount "traefik-staging-dask-gateway"
client.go:684: [debug] Looks like there are no changes for ServiceAccount "hub"
client.go:684: [debug] Looks like there are no changes for ServiceAccount "user-scheduler"
client.go:684: [debug] Looks like there are no changes for ServiceAccount "user-sa"
client.go:693: [debug] Patch Secret "hub" in namespace staging
client.go:693: [debug] Patch Secret "singleuser" in namespace staging
client.go:684: [debug] Looks like there are no changes for ConfigMap "controller-staging-dask-gateway"
client.go:693: [debug] Patch ConfigMap "api-staging-dask-gateway" in namespace staging
client.go:684: [debug] Looks like there are no changes for ConfigMap "hub"
client.go:684: [debug] Looks like there are no changes for ConfigMap "user-scheduler"
client.go:684: [debug] Looks like there are no changes for ConfigMap "basehub-cluster-info"
client.go:684: [debug] Looks like there are no changes for ConfigMap "hub-custom-templates-config"
client.go:684: [debug] Looks like there are no changes for PersistentVolume "staging-home-nfs"
client.go:684: [debug] Looks like there are no changes for PersistentVolumeClaim "hub-db-dir"
client.go:684: [debug] Looks like there are no changes for PersistentVolumeClaim "home-nfs"
client.go:684: [debug] Looks like there are no changes for ClusterRole "controller-staging-dask-gateway"
client.go:684: [debug] Looks like there are no changes for ClusterRole "api-staging-dask-gateway"
client.go:684: [debug] Looks like there are no changes for ClusterRole "traefik-staging-dask-gateway"
client.go:684: [debug] Looks like there are no changes for ClusterRole "staging-user-scheduler"
client.go:684: [debug] Looks like there are no changes for ClusterRoleBinding "controller-staging-dask-gateway"
client.go:684: [debug] Looks like there are no changes for ClusterRoleBinding "api-staging-dask-gateway"
client.go:684: [debug] Looks like there are no changes for ClusterRoleBinding "traefik-staging-dask-gateway"
client.go:684: [debug] Looks like there are no changes for ClusterRoleBinding "staging-user-scheduler"
client.go:684: [debug] Looks like there are no changes for Role "hub"
client.go:684: [debug] Looks like there are no changes for RoleBinding "hub"
client.go:684: [debug] Looks like there are no changes for Service "api-staging-dask-gateway"
client.go:684: [debug] Looks like there are no changes for Service "traefik-staging-dask-gateway"
client.go:684: [debug] Looks like there are no changes for Service "hub"
client.go:684: [debug] Looks like there are no changes for Service "proxy-api"
client.go:684: [debug] Looks like there are no changes for Service "proxy-public"
client.go:684: [debug] Looks like there are no changes for Service "configurator"
client.go:693: [debug] Patch Deployment "controller-staging-dask-gateway" in namespace staging
client.go:693: [debug] Patch Deployment "api-staging-dask-gateway" in namespace staging
client.go:693: [debug] Patch Deployment "traefik-staging-dask-gateway" in namespace staging
client.go:693: [debug] Patch Deployment "hub" in namespace staging
client.go:693: [debug] Patch Deployment "proxy" in namespace staging
client.go:693: [debug] Patch Deployment "user-scheduler" in namespace staging
client.go:693: [debug] Patch Deployment "shared-dirsize-metrics" in namespace staging
client.go:684: [debug] Looks like there are no changes for Deployment "shared-volume-metrics"
client.go:693: [debug] Patch StatefulSet "user-placeholder" in namespace staging
client.go:684: [debug] Looks like there are no changes for Ingress "jupyterhub"
client.go:693: [debug] Patch IngressRoute "api-staging-dask-gateway" in namespace staging
client.go:693: [debug] Patch Middleware "clusters-prefix-staging-dask-gateway" in namespace staging
client.go:693: [debug] Patch Middleware "api-prefix-staging-dask-gateway" in namespace staging
upgrade.go:428: [debug] waiting for release staging resources (created: 0 updated: 53  deleted: 0)
wait.go:48: [debug] beginning wait for 53 resources with timeout of 5m0s
ready.go:303: [debug] Deployment is not ready: staging/api-staging-dask-gateway. 0 out of 1 expected pods are ready
ready.go:303: [debug] Deployment is not ready: staging/api-staging-dask-gateway. 0 out of 1 expected pods are ready
ready.go:303: [debug] Deployment is not ready: staging/api-staging-dask-gateway. 0 out of 1 expected pods are ready
ready.go:303: [debug] Deployment is not ready: staging/api-staging-dask-gateway. 0 out of 1 expected pods are ready
ready.go:303: [debug] Deployment is not ready: staging/api-staging-dask-gateway. 0 out of 1 expected pods are ready
ready.go:436: [debug] StatefulSet is ready: staging/user-placeholder. 0 out of 0 expected pods are ready
upgrade.go:171: [debug] updating status for upgraded release for staging
```

That's a bunch of changes! So I just ran it again, which should ideally prouce no changes. yet...

```
history.go:56: [debug] getting history for release staging
upgrade.go:155: [debug] preparing upgrade for staging
upgrade.go:163: [debug] performing update for staging
upgrade.go:356: [debug] creating upgraded release for staging
client.go:486: [debug] Starting delete for "nfs-share-creator" Job
client.go:490: [debug] Ignoring delete failure for "nfs-share-creator" batch/v1, Kind=Job: jobs.batch "nfs-share-creator" not found
wait.go:66: [debug] beginning wait for 1 resources to be deleted with timeout of 5m0s
client.go:142: [debug] creating 1 resource(s)
client.go:712: [debug] Watching for changes to Job nfs-share-creator with timeout of 5m0s
client.go:740: [debug] Add/Modify event for nfs-share-creator: ADDED
client.go:779: [debug] nfs-share-creator: Jobs active: 1, jobs failed: 0, jobs succeeded: 0
client.go:740: [debug] Add/Modify event for nfs-share-creator: MODIFIED
client.go:779: [debug] nfs-share-creator: Jobs active: 0, jobs failed: 0, jobs succeeded: 0
client.go:740: [debug] Add/Modify event for nfs-share-creator: MODIFIED
client.go:486: [debug] Starting delete for "nfs-share-creator" Job
wait.go:66: [debug] beginning wait for 1 resources to be deleted with timeout of 5m0s
client.go:393: [debug] checking 53 resources for changes
client.go:693: [debug] Patch PriorityClass "staging-default-priority" in namespace 
client.go:693: [debug] Patch PriorityClass "staging-user-placeholder-priority" in namespace 
client.go:693: [debug] Patch NetworkPolicy "hub" in namespace staging
client.go:693: [debug] Patch NetworkPolicy "proxy" in namespace staging
client.go:693: [debug] Patch NetworkPolicy "singleuser" in namespace staging
client.go:693: [debug] Patch PodDisruptionBudget "user-placeholder" in namespace staging
client.go:693: [debug] Patch PodDisruptionBudget "user-scheduler" in namespace staging
client.go:684: [debug] Looks like there are no changes for ServiceAccount "controller-staging-dask-gateway"
client.go:684: [debug] Looks like there are no changes for ServiceAccount "api-staging-dask-gateway"
client.go:684: [debug] Looks like there are no changes for ServiceAccount "traefik-staging-dask-gateway"
client.go:684: [debug] Looks like there are no changes for ServiceAccount "hub"
client.go:684: [debug] Looks like there are no changes for ServiceAccount "user-scheduler"
client.go:684: [debug] Looks like there are no changes for ServiceAccount "user-sa"
client.go:693: [debug] Patch Secret "hub" in namespace staging
client.go:693: [debug] Patch Secret "singleuser" in namespace staging
client.go:684: [debug] Looks like there are no changes for ConfigMap "controller-staging-dask-gateway"
client.go:684: [debug] Looks like there are no changes for ConfigMap "api-staging-dask-gateway"
client.go:684: [debug] Looks like there are no changes for ConfigMap "hub"
client.go:684: [debug] Looks like there are no changes for ConfigMap "user-scheduler"
client.go:684: [debug] Looks like there are no changes for ConfigMap "basehub-cluster-info"
client.go:684: [debug] Looks like there are no changes for ConfigMap "hub-custom-templates-config"
client.go:684: [debug] Looks like there are no changes for PersistentVolume "staging-home-nfs"
client.go:684: [debug] Looks like there are no changes for PersistentVolumeClaim "hub-db-dir"
client.go:684: [debug] Looks like there are no changes for PersistentVolumeClaim "home-nfs"
client.go:684: [debug] Looks like there are no changes for ClusterRole "controller-staging-dask-gateway"
client.go:684: [debug] Looks like there are no changes for ClusterRole "api-staging-dask-gateway"
client.go:684: [debug] Looks like there are no changes for ClusterRole "traefik-staging-dask-gateway"
client.go:684: [debug] Looks like there are no changes for ClusterRole "staging-user-scheduler"
client.go:684: [debug] Looks like there are no changes for ClusterRoleBinding "controller-staging-dask-gateway"
client.go:684: [debug] Looks like there are no changes for ClusterRoleBinding "api-staging-dask-gateway"
client.go:684: [debug] Looks like there are no changes for ClusterRoleBinding "traefik-staging-dask-gateway"
client.go:684: [debug] Looks like there are no changes for ClusterRoleBinding "staging-user-scheduler"
client.go:684: [debug] Looks like there are no changes for Role "hub"
client.go:684: [debug] Looks like there are no changes for RoleBinding "hub"
client.go:684: [debug] Looks like there are no changes for Service "api-staging-dask-gateway"
client.go:684: [debug] Looks like there are no changes for Service "traefik-staging-dask-gateway"
client.go:684: [debug] Looks like there are no changes for Service "hub"
client.go:684: [debug] Looks like there are no changes for Service "proxy-api"
client.go:684: [debug] Looks like there are no changes for Service "proxy-public"
client.go:684: [debug] Looks like there are no changes for Service "configurator"
client.go:693: [debug] Patch Deployment "controller-staging-dask-gateway" in namespace staging
client.go:693: [debug] Patch Deployment "api-staging-dask-gateway" in namespace staging
client.go:693: [debug] Patch Deployment "traefik-staging-dask-gateway" in namespace staging
client.go:693: [debug] Patch Deployment "hub" in namespace staging
client.go:693: [debug] Patch Deployment "proxy" in namespace staging
client.go:693: [debug] Patch Deployment "user-scheduler" in namespace staging
client.go:693: [debug] Patch Deployment "shared-dirsize-metrics" in namespace staging
client.go:684: [debug] Looks like there are no changes for Deployment "shared-volume-metrics"
client.go:693: [debug] Patch StatefulSet "user-placeholder" in namespace staging
client.go:684: [debug] Looks like there are no changes for Ingress "jupyterhub"
client.go:693: [debug] Patch IngressRoute "api-staging-dask-gateway" in namespace staging
client.go:693: [debug] Patch Middleware "clusters-prefix-staging-dask-gateway" in namespace staging
client.go:693: [debug] Patch Middleware "api-prefix-staging-dask-gateway" in namespace staging
upgrade.go:428: [debug] waiting for release staging resources (created: 0 updated: 53  deleted: 0)
wait.go:48: [debug] beginning wait for 53 resources with timeout of 5m0s
ready.go:436: [debug] StatefulSet is ready: staging/user-placeholder. 0 out of 0 expected pods are ready
upgrade.go:171: [debug] updating status for upgraded release for staging
```

ok, so that's great - those are unrelated to this PR.

I spot checked by looking at the hub deployment object to see what has changed between deploys. The answer is nothing, so we're good from that respect.

yuvipanda commented:
Then I collected output of `--debug` doing a deployment from `main`, and then from this branch. I looked through the diff. Here are some of the differences found:

1. No differences in content, but the source has changed. This is expected. example:
```
-# Source: daskhub/charts/dask-gateway/templates/traefik/rbac.yaml
+# Source: daskhub/charts/basehub/charts/dask-gateway/templates/traefik/rbac.yaml
 kind: ServiceAccount
 apiVersion: v1
 metadata:
   name: traefik-staging-dask-gateway
```

2. Some were no-ops, but described as removals and re-creations. This is almost entirely due to ordering differences in the `helm upgrade` runs and a result of lack of sorting. can be ignored.

```
-# Source: daskhub/charts/basehub/charts/jupyterhub/templates/hub/serviceaccount.yaml
-apiVersion: v1
-kind: ServiceAccount
-metadata:
-  name: hub
-  labels:
-    component: hub
-    app: jupyterhub
-    release: staging
-    chart: jupyterhub-3.3.7
-    heritage: Helm
---
+# Source: daskhub/charts/basehub/charts/jupyterhub/templates/hub/serviceaccount.yaml
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: hub
+  labels:
+    component: hub
+    app: jupyterhub
+    release: staging
+    chart: jupyterhub-3.3.7
+    heritage: Helm

````

Not sure what's up with these, but harmless in this context.

3. Change to the *contents* of the `values.yaml`. This is to be expected, and is harmless by itself.

That's it!

It's still helpful to do these tests (described in https://github.com/2i2c-org/infrastructure/pull/4240#issuecomment-2195879982) as you deploy these @GeorgianaElena but this looks great!

GeorgianaElena commented:
Thanks @yuvipanda <3

I've checked and we do have a couple of daskhub tests similar with the one you shared at https://github.com/2i2c-org/infrastructure/blob/4db4e3255ec31ebe4affb1ffb2636c6d4d9d78bb/deployer/health_check_tests/test-notebooks/daskhub/scale_dask_workers.ipynb

I pushed one commit that should make them run correctly for the new dask-gateaway setup. I also ran them locally and made sure a dask worker pod gets created, then destroyed. So I'll go ahead and trust these tests, go for a merge and investigate if there are any health checks that fail.
